feat: coroutine flows documentation rewrite (#4687)
Kotlin/kotlinx.coroutines: docs/topics/coroutines-flow.md

A flow pipeline is a sequence of operations that involves the following roles:

* **Emitter**: produces values.
* **Intermediate operators (optional)**: consume values from a flow, apply an operation to them, and return another flow.
* **Collector**: consumes values from a flow.

Here's a simple example that shows how these pipeline roles work together:

```kotlin
import kotlinx.coroutines.flow.*

//sampleStart
suspend fun main() {
    // The emitter produces values
    flowOf(0x4B, 0x6F, 0x74, 0x6C, 0x69, 0x6E)
        // The intermediate operator consumes values,
        // applies an operation, and returns another flow
        .map { value -> value.toChar() }
        // The collector consumes the transformed values
        .collect { updatedValue ->
            println("Say '$updatedValue'!")
        }
}
//sampleEnd
```
{kotlin-runnable="true"}

In a flow, values move from an emitter toward a collector, from *upstream* to *downstream*.
Intermediate operators collect an upstream flow, apply an operation to its values, and return a new downstream flow.
That downstream flow can become the upstream flow for the next collector.

> ![Parts of a flow: emitter, intermediate operator (optional), collector. Values move from upstream to downstream.](flow-upstream-downstream.svg){width="700px"}

Reviewer: ```suggestion
![Parts of a flow: emitter, intermediate operator (optional), collector. Values move from upstream to downstream.](flow-upstream-downstream.svg){width=700}
```

Reviewer: nitpick: the upstream/downstream line at the bottom is not aligned with the text bubbles above

Author: ah cheers — I aligned it 👍

Kotlin provides the following flow types:

* [**Cold flows**](#cold-flows) start producing values when collected.
  Each collector triggers a new, independent execution of the flow.
* [**Hot flows**](#hot-flows) emit values independently of collectors and share the same stream of values with all collectors.

> You can use the [Turbine library](https://github.com/cashapp/turbine) to test Kotlin flows.
> It simplifies collecting and asserting flow emissions in unit tests, including completion and failure cases.
> 
{style="tip"}

## Cold flows

> Like [sequences](sequences.md), cold flows are lazy.

Reviewer: maybe a link to an explanation about 'lazy'?

Author: Do we have one? I think since it's a generic programming concept it never made it into the docs/ I never managed to find a generic one (we have descriptions for lazy properties and lazy init in Kotlin/JS 😢  — we have some APIs where it's described in the given context 🤔 I think it should be fine, the next sentence explains immediately what it means in this context.

The code block of a cold flow builder doesn't run until a collector collects it.
Each new collector starts a new execution of the flow.

### Create a cold flow

To create a cold flow, use the [`flow()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/flow.html) builder function.
Inside its block, use the [`emit()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-flow-collector/emit.html) function to emit values to collectors:

```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

//sampleStart
fun main() {
    // Creates a flow
    val pageFlow = flow {
        for (page in 1..3) {
            println("Loading page $page...")

            // Emits each page as it is loaded
            emit("Page $page")
        }
    }
    println("Creating a cold flow doesn't run it!")
}
//sampleEnd
```
{kotlin-runnable="true"}

In this example, the `flow()` builder function returns a [`Flow<T>`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-flow/), but it doesn't start executing its block.

> A cold flow is like a recipe: it defines how to produce values, but it only starts producing values when you [collect it](#collect-from-a-cold-flow).

Reviewer: ```suggestion
A cold flow is like a recipe: it defines how to produce values, but it only starts producing values when you [collect it](#collect-a-cold-flow).
```

You can also create cold flows with the following functions:

* [`flowOf()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/flow-of.html): creates a flow from provided values.
* [`.asFlow()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/as-flow.html): converts an existing iterable, such as a range, into a flow.

Here's an example:

```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

//sampleStart
fun main() {
    // Creates a flow from provided values
    val predefinedPageFlow = flowOf("Page 1", "Page 2", "Page 3")
    // Creates a flow from a range
    val generatedPageFlow = (1..3).asFlow()
}
//sampleEnd
```

> {kotlin-runnable="true"}

Reviewer: not sure if this one should be runnable, no output

Author: shouldn't be 👍 I removed runnable.

        println("Starting emissions in $coroutineName")
        for (page in 1..3) {
            println("Loading page $page in $coroutineName")
            emit("Page $page")
        }
        println("Done emitting in $coroutineName")
    }

    withContext(Dispatchers.Default) {
        // Launches a collector that processes each page slowly
        launch(CoroutineName("a slow coroutine")) {
            pageFlow.collect {
                println("Processing $it slowly")
                delay(100.milliseconds)
                println("Done processing $it slowly")
            }
        }

        // Launches a collector that processes each page quickly
        launch(CoroutineName("a fast coroutine")) {
            pageFlow.collect {
                println("Processing $it quickly")
                delay(10.milliseconds)
                println("Done processing $it quickly")
            }
        }
    }
}
//sampleEnd
```
{kotlin-runnable="true"}

In this example, [`CoroutineName`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-name/) adds a name to each coroutine.

> You can use `CoroutineName` for [debugging](coroutine-context-and-dispatchers.md#naming-coroutines-for-debugging), and here it helps show which collector runs each collection.

Reviewer: ```suggestion
You can use `CoroutineName` for [debugging](coroutine-context-and-dispatchers.md#naming-coroutines-for-debugging). Here, it helps show which collector runs each collection.
```

### Intermediate flow operators

Intermediate operators apply operations to an upstream flow and return a new downstream flow.
They are cold, so the returned flow doesn't start processing values until it's collected, even when the upstream flow is hot.

> The [kotlinx.coroutines](https://github.com/Kotlin/kotlinx.coroutines) library offers a [wide range of intermediate flow operators](coroutines-flow-operators.md) for transforming and processing flows.

Reviewer: I think the link to the operators page should be commented out for now?

Author: Based on the discussion in the channel, we'll add this page, but only with the current content (without any updates)   — I'll add it when pushing the changes to fix the link issue

Reviewer: then I think we should add What's next to that 2nd page at the bottom

Author: Yes definitely, but maybe only add it when we update it (since that one will be fully rewritten) WDYT?

### Handle exceptions in flows

Both emitters and collectors can throw exceptions.

If you don't handle an exception during flow collection, it propagates upstream from the collector and is thrown to the caller of the `collect()` function.

You can handle such exceptions by wrapping the `collect()` function in a `try-catch` block, for example:

```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

class MyFlowException(message: String) : Exception(message)

//sampleStart
suspend fun main() {
    val myFlow = flow {
        try {
            // The emit() function calls the lambda passed to collect()
            emit('a')
        } catch (e: MyFlowException) {
            println("Collector threw $e")

            // Rethrows the downstream exception
            throw e
        }
    }
    // Wraps flow collection in try-catch
    try {
        myFlow.collect {
            // Throws an exception from the collect() lambda
            throw MyFlowException("Can't process '$it'!")
        }
    } catch (e: MyFlowException) {
        println("Flow collection failed with $e")
        // Rethrows the exception to the caller
        throw e
    }
}
//sampleEnd
```

> {kotlin-runnable="true"}

Reviewer: here I only get "Cannot convert undefined or null to object"

Author: That's weird 🤔 I get: 

Collector threw MyFlowException: Can't process 'a'!
Flow collection failed with MyFlowException: Can't process 'a'!
Exception in thread "main" MyFlowException: Can't process 'a'!
 at FileKt$main$2.emit (File.kt:22) 
 at FileKt$main$2.emit (File.kt:20) 
 at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke (SafeCollector.kt:15)

Here's an example where a flow loads data and reports its progress:

```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlin.random.*
import java.io.IOException
import kotlin.time.Duration.Companion.milliseconds

//sampleStart
sealed interface LoadingState {
    sealed interface Terminal: LoadingState
    object Started: LoadingState
    data class Percentage(val percents: Int): LoadingState
    object Failed: Terminal
    object Done: Terminal
}

fun loadBlob(url: String) = flow {
    emit(LoadingState.Started)
    repeat(100) {
        if (Random.nextInt(100) == 13)
            throw IOException("Failed to load!")
        emit(LoadingState.Percentage(it))
        delay(10.milliseconds)
    }
    emit(LoadingState.Done)
}.catch { e ->
    println("Loading data failed with $e")
    if (e is IOException) {
        // Handles an expected exception
        emit(LoadingState.Failed)
    } else {
        // Rethrows unexpected exceptions, so the collect() fails with them
        throw e
    }
}

suspend fun main() {
    loadBlob("https://example.com/").collect {
        println("Got '$it'")
    }
}
//sampleEnd
```

> {kotlin-runnable="true"}

Reviewer: Is it expected to get all 100% or not? (getting "Loading data failed with java.io.IOException: Failed to load!
Got 'LoadingState$Failed@7466d320' after 84%)

Author: nope not always. that's why we handle it as an expected exception here: 

```
.catch { e ->
    println("Loading data failed with $e")
    if (e is IOException) {
        // Handles an expected exception
        emit(LoadingState.Failed)
    } else {
        // Rethrows unexpected exceptions, so the collect() fails with them
        throw e
    }
```

        println("Flow collection failed with $e")
    }
}
```
{kotlin-runnable="true"}

Since the `collect()` lambda runs after `.catch()`, you can't use `.catch()` to handle exceptions thrown from it.
To handle exceptions from code that runs for each emitted value with `.catch()`, place that code in [.onEach()](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/on-each.html) before `.catch()`.

The `.onEach()` operator runs its lambda before each value is emitted downstream.
If `.catch()` handles an exception from `.onEach()`, the flow completes and doesn't emit the next value:

Here's an example:

```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlin.random.*
import java.io.IOException
import kotlin.time.Duration.Companion.milliseconds

suspend fun main() {
    flowOf('a', 'o', '5', 'c')
        // Runs before each value is emitted downstream
        .onEach {
            require(!it.isDigit()) { "Digits are not allowed!" }
            println("Got '$it'")
        }
        .catch { e ->
            println("Caught an exception: $e")
        }
        .collect()
}
```

> {kotlin-runnable="true"}

Reviewer: MismatchedInputException error

Author: I can't reproduce it 😢

In this example, the `.onEach()` operator runs upstream from `.catch()`, so the `.catch()` operator handles the exception when the `require()` check fails for `'5'`.

#### Restart the upstream flow after an exception

> Some operations may fail temporarily, such as a network request that loses connection.

Reviewer: ```suggestion
Some operations may temporarily fail, such as a network request that loses its connection.
```

import kotlin.time.Duration.Companion.seconds

//sampleStart
sealed interface LoadingState {
    sealed interface Terminal: LoadingState
    object Started: LoadingState
    data class Percentage(val percents: Int): LoadingState
    object Failed: Terminal
    object Done: Terminal
}

fun loadBlob(url: String) = flow {
    emit(LoadingState.Started)
    repeat(100) {
        if (Random.nextInt(100) == 13)
            throw IOException("Failed to load!")
        emit(LoadingState.Percentage(it))
        delay(10.milliseconds)
    }
    emit(LoadingState.Done)
}.retry(3) { e ->
    if (e is IOException) {
        // This is an expected error
        // Waits for one second before retrying
        delay(1.seconds)
        true
    } else {
        // Stops retrying and rethrows unexpected exceptions
        false
    }
}

suspend fun main() {
    loadBlob("https://example.org/").collect {
        println("Got $it")
    }
}
//sampleEnd
```
{kotlin-runnable="true"}

### Flow cancellation

Flow cancellation stops collection when the result is no longer needed, such as when a request times out.

Flow collection is tied to the coroutine that calls the `collect()` function.

> When that coroutine is canceled, collection stops and the upstream flow is canceled too.

Reviewer: ```suggestion
When that coroutine is canceled, the collection stops, and the upstream flow is canceled too.
```

To cancel flow collection, call the `cancel()` function on the `Job` of the collecting coroutine:

```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlin.random.*
import java.io.IOException
import kotlin.time.Duration.Companion.milliseconds

//sampleStart
val myFlow = flow {
    var i = 0
    try {
        while (true) {
            println("Emitting $i")
            emit(i)
            println("Emitted $i")
            ++i
            delay(10.milliseconds)
        }
    } catch (e: Throwable) {
        println("Upstream finished with $e")
        throw e
    }
}

suspend fun main() {
    coroutineScope {
        val job = launch {
            try {
                myFlow.collect {
                    println("Processing $it")
                    delay(5.milliseconds)
                }
            } catch (e: Throwable) {
                println("Collection finished with $e")
                throw e
            }
        }
        delay(100.milliseconds)

        // Cancels the coroutine that collects the flow
        job.cancel()
    }
}
//sampleEnd
```

> {kotlin-runnable="true"}

Reviewer: "Cannot convert undefined or null to object" error

Author: I can't reproduce this one either 😢 works for me

```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlin.random.*
import java.io.IOException
import kotlin.time.Duration.Companion.milliseconds

//sampleStart
// Defines a simplified version of the default .take() operator
fun <T> Flow<T>.myTake(count: Int): Flow<T> = flow {
    require(count > 0)
    val cancellationException = CancellationException()
    var elementsRemaining = count
    try {
        this@myTake.collect {
            emit(it)
            --elementsRemaining
            if (elementsRemaining == 0) {
                // Cancels the upstream flow after the requested number of values
                throw cancellationException
            }
        }
    } catch (e: Throwable) {
        if (e === cancellationException) {
            // Handles the CancellationException used to cancel the upstream flow
            // Completes the flow after the set number of values in .myTake()
        } else {
            // Rethrows unexpected exceptions
            throw e
        }
    }
}

suspend fun main() {
    (0..1000).asFlow().myTake(3).collect {
        println("Got $it")
    }
}
//sampleEnd
```

> {kotlin-runnable="true"}

Reviewer: java.io.IOException: Cannot run program

Author: same 😢 works for me.

> In this example, the `.myTake()` function emits values from the upstream flow until it emits the requested number of values.

Reviewer: ```suggestion
In this example, the `.myTake()` function emits values from the upstream flow until it emits the requested number.
```

Author: I think that's a bit less specific 🤔 (I'm worried it could be misread as `.take(2)` means until it emits the number `2` not two values.

Author: so perhaps let's change it to specified number of values to be more clear? 🤔

Reviewer: or "until all requested values are emitted."

Author: that works as well 👍 although it's passive, but I think it's understandable.

```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlin.random.*
import java.io.IOException
import kotlin.time.Duration.Companion.milliseconds

//sampleStart
// Defines a simplified version of the default .merge() operator
fun <T> Flow<T>.myMerge(other: Flow<T>): Flow<T> = channelFlow {
    // CoroutineScope and SendChannel are available as receivers here
    // Launches a coroutine that collects the receiver flow
    launch {
        // Collects the receiver flow
        this@myMerge.collect {
            send(it)
        }
    }
    launch {
        // Launches a coroutine that collects the other flow
        other.collect {
            // Calls SendChannel.send
            send(it)
        }
    }
}

suspend fun main() {
    val flow1 = (0..3).asFlow().onEach { delay(20.milliseconds) }
    val flow2 = (6..9).asFlow().onEach { delay(50.milliseconds) }
    flow1.myMerge(flow2).collect { println(it) }
}
//sampleEnd
```

> {kotlin-runnable="true"}

Reviewer: Cannot convert undefined or null to object

Author: same — works for me 😢

When the buffer is full, producers suspend until the buffer has free space.

You can change the buffer capacity with the [`.buffer()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/buffer.html) operator.
For example, `.buffer(12)` lets producers send up to 12 values ahead of the collector, while `.buffer(0)` removes the buffer, so each value is sent only when the collector can receive it.

Here's an example:

```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlin.random.*
import java.io.IOException
import kotlin.time.Duration.Companion.milliseconds

//sampleStart
suspend fun main() {
    val oneHundredNumbers = channelFlow {
        repeat(100) {
            println("Sending $it")
            send(it)
        }
    }

    // Uses the default buffer capacity
    oneHundredNumbers.collect {
        println("Processing $it")
        delay(10.milliseconds)
    }
  
    // Removes the buffer so sending and processing interleave from the start
    oneHundredNumbers.buffer(0).collect {
        println("Processing $it")
        delay(10.milliseconds)
    }
}
//sampleEnd
```

> {kotlin-runnable="true"}

Reviewer: Cannot convert undefined or null to object

Author: same — should work 😢

Reviewer: It works now, maybe 100 is too many? the output is so long on the page

Author: good point, let's go in 10 increments -- 10..20..30.. etc. I'll change it

With the default buffer capacity, the producer sends values quickly until the buffer becomes full.
After that, `send()` suspends until the buffer has free space, so `Sending` and `Processing` messages start to interleave.

With `.buffer(0)`, each `send()` call waits until the collector can receive the value, so `Sending` and `Processing` interleave from the start.

## Hot flows

Hot flows are shared streams that emit values independently of collectors.
They keep emitting values even when no collector is active, and multiple collectors can collect the same emissions from an already active stream instead of starting a new execution.

A collector of a hot flow is called a *subscriber*.

You can use hot flows when multiple parts of an application need to react to the same stream of updates, such as incoming chat messages, user actions, or UI state changes.

Kotlin provides two hot flow types:

* [`SharedFlow`](#create-a-sharedflow) broadcasts values to multiple subscribers. Use it when you need to broadcast events that happen over time, such as messages or notifications.

> * [`StateFlow`](#create-a-stateflow) is a specialized `SharedFlow` that always holds the latest state value. Use it when you need to represent state that changes over time, such as UI state.

Reviewer: ```suggestion
* [`StateFlow`](#create-a-stateflow) is a specialized `SharedFlow` that always holds the latest state value. Use it when you need to represent a state that changes over time, such as UI state.
```

Author: 🤔 here, state refers to uncountable concepts so I'd go without an article

### Create a `SharedFlow`

[`SharedFlow`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-shared-flow/) is a hot flow that broadcasts emitted values that happen over time to its subscribers.

> You can create `SharedFlow` with the [`MutableSharedFlow()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-mutable-shared-flow.html) function.

Reviewer: ```suggestion
You can create a `SharedFlow` with the [`MutableSharedFlow()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-mutable-shared-flow.html) function.
```

A `MutableSharedFlow` exposes functions for emitting values.
If you expose it directly, code outside the class can emit values into the flow.

> To prevent this, store the mutable flow in a private [backing property](properties.md#backing-properties), and expose a read-only `SharedFlow` with the [`.asSharedFlow()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/as-shared-flow.html) function.

Reviewer: ```suggestion
To prevent this, store the mutable flow in a private [backing property](properties.md#backing-properties) and expose a read-only `SharedFlow` with the [`.asSharedFlow()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/as-shared-flow.html) function.
```

Author: hmm 🤔 I prefer without a comma here because it's a compound predicate

Reviewer: I don't think so, it's basically 2 coordinated verbs in the main clause "store and expose"

Author: Oh I get what you mean — I agree 👍

To emit values to subscribers, use the `emit()` function on the `MutableSharedFlow`:

```kotlin
data class Message(
    val senderId: Int,
    val time: Instant,
    val text: String,
)

class Chatroom {
    // Stores the SharedFlow in a private backing property
    private val _messages = MutableSharedFlow<Message>()

    // Exposes a read-only SharedFlow to subscribers
    val messages: SharedFlow<Message>
        get() = _messages.asSharedFlow()

    suspend fun sendMessageToEveryone(message: Message) {
        // Emits the message to subscribers
        _messages.emit(message)
    }
}
```

Just like with cold flows, you can use the `collect()` function to collect values from a `SharedFlow`.

You can also configure a `SharedFlow` to immediately replay already emitted values to new subscribers.

> The replay cache works like a small history buffer and stores a fixed number of previous emissions.

Reviewer: ```suggestion
The replay cache works like a small history buffer, storing a fixed number of previous emissions.
```

To set how many previous emissions a new subscriber receives, use the `replay` parameter in `MutableSharedFlow()`:

```kotlin

> // Sets the number of already emitted messages new subscribers receive on subscription

Reviewer: ```suggestion
// Sets the number of already emitted messages that new subscribers receive on subscription
```

const val MESSAGES_TO_REMEMBER = 10

class Chatroom {
    private val _messages = MutableSharedFlow<Message>(

        // Replays the set amount of last emitted messages to new subscribers
        replay = MESSAGES_TO_REMEMBER
    )

    val messages: SharedFlow<Message>
        get() = _messages.asSharedFlow()

    suspend fun sendMessageToEveryone(message: Message) {

>         // Emits the message to subscribes of the messages flow

Reviewer: ```suggestion
        // Emits the message to subscribers of the message flow
```

Author: ah nice catch on the typo! 🙏 — the flow should be messages (message is a single emitted value)

        _messages.emit(message)
    }
}
```

Collecting a hot flow doesn't complete by itself, so you must [cancel the collecting coroutines](#cancel-hot-flows) when you no longer need them.

> A hot flow doesn't have a close or cancel operation.

> > Canceling collection only stops that subscriber from collecting.

Reviewer: ```suggestion
> Canceling a collection only stops that subscriber from collecting.
```

Author: hmm it should be again a concept here, but I see what could be the issue (stops that subscriber makes it read a bit like we are talking about a specific situation and not a general one) 🤔 

I rewrote it as: 

> Canceling collection only stops the corresponding subscriber from collecting.

    // Emits the message to subscribers
    suspend fun sendMessageToEveryone(message: Message) {
        _messages.emit(message)
    }
}

suspend fun main() {
    val nUsers = 3
    val chatroom = Chatroom()
    withContext(Dispatchers.Default) {
        // Starts a message reader for each user
        val messageReaders = List(nUsers) { userId ->
            // Starts collection before messages are emitted
            launch(start = CoroutineStart.UNDISPATCHED) {
                chatroom.messages.collect { message ->
                    println("User $userId received $message")
                }
            }
        }
        // Sends a greeting from each user
        repeat(nUsers) { userId ->
            chatroom.sendMessageToEveryone(
                Message(
                    userId,
                    Clock.System.now(),
                    "Hello from $userId!"
                )
            )
        }
        // Delays to make sure people have enough time to chat
        delay(100.milliseconds)
        // Cancels readers because SharedFlow collection doesn't finish by itself
        messageReaders.forEach { it.cancel() }
    }

}
```

> {kotlin-runnable="true"}

Reviewer: Cannot convert undefined or null to object

In this example, [`CoroutineStart.UNDISPATCHED`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-start/-u-n-d-i-s-p-a-t-c-h-e-d/) starts each collecting coroutine immediately.

This ensures that each coroutine reaches `collect()`, subscribes to `messages`, and suspends before `sendMessageToEveryone()` emits messages.
Without it, a collecting coroutine might start later and miss earlier emissions if the replay cache is too small.

#### Use explicit backing fields to expose hot flows
<primary-label ref="experimental-opt-in"/>

You can use [explicit backing fields](whatsnew23.md#explicit-backing-fields) to expose a read-only `SharedFlow` while keeping a mutable backing field inside the class.

Explicit backing fields define the implementation type in the `field` declaration.
Inside the class, the compiler smart casts the property to the backing field type, so you can call the `emit()` function without a separate private backing property.

> > Explicit backing fields doesn't create the read-only wrapper that `.asSharedFlow()` provides.

Reviewer: ```suggestion
> Explicit backing fields don't create the read-only wrapper that `.asSharedFlow()` provides.
```

    }
}
//sampleEnd

suspend fun main() {
    val nUsers = 3
    val chatroom = Chatroom()

    withContext(Dispatchers.Default) {
        val messageReaders = List(nUsers) { userId ->
            launch(start = CoroutineStart.UNDISPATCHED) {
                chatroom.messages.collect { message ->
                    println("User $userId received $message")
                }
            }
        }

        repeat(nUsers) { userId ->
            chatroom.sendMessageToEveryone(
                Message(
                    senderId = userId,
                    time = Clock.System.now(),
                    text = "Hello from $userId!"
                )
            )
        }

        delay(100.milliseconds)
        messageReaders.forEach { it.cancel() }
    }
}
```
{kotlin-runnable="true"}

### Create a `StateFlow`

[`StateFlow`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-state-flow/) is a hot flow that stores a single state value and emits updates when that value is replaced with a new one.
New subscribers receive the current value as soon as they start collecting, and then receive new values each time the state updates.

> You can use `StateFlow` to represent state that changes over time, such as loading progress, UI state, or the state of an object.

Reviewer: ```suggestion
You can use `StateFlow` to represent a state that changes over time, such as loading progress, UI state, or the state of an object.
```

Author: same as above - should be an uncountable state

val result = MutableStateFlow<LoadingState>(LoadingState.Started)
```

To set the current state, use the [`value`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-mutable-state-flow/value.html) property:

```kotlin
fun loadBlob(url: String): StateFlow<LoadingState> {
    val result = MutableStateFlow<LoadingState>(LoadingState.Started)

    DownloadManager.startLoading(
        url,
        onPercentageLoaded = { percentage ->
            // Replaces the current state with the latest progress
            result.value = LoadingState.Percentage(percentage)
        },
        onCompletion = {
            // Replaces the current state with the completion state
            result.value = LoadingState.Done
        },
        onFailure = {
            // Replaces the current state with the failure state
            result.value = LoadingState.Failed
        }
    )
}
```

> Setting `value` is thread-safe and replaces the current state, but updating `value` based on its previous value isn't atomic.
> When a new state depends on the previous state, use `.update()` instead.
>
{style="note"}

> Similarly to `MutableSharedFlow`, `MutableStateFlow` exposes APIs for emitting updates.

Reviewer: ```suggestion
Similar to `MutableSharedFlow`, `MutableStateFlow` exposes APIs for emitting updates.
```

                    return@launch
                }
                onPercentageLoaded(it)
                delay(10.milliseconds)
            }
            onCompletion()
        }
    }
}

suspend fun main() {
    loadBlob("https://example.com/").onEach { state ->
        when (state) {
            is LoadingState.Started -> {
                // Waits for progress updates
            }
            is LoadingState.Percentage ->
                println("Loaded ${state.percents}...")
            is LoadingState.Failed ->
                println("Loading failed.")
            is LoadingState.Done ->
                println("Finished loading!")
        }
    }.takeWhile { it !is LoadingState.Terminal }.collect()
}
//sampleEnd
```
{kotlin-runnable="true"}

> This example uses [`GlobalScope`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-global-scope/) only to keep the callback-based API short.
> In your own applications, pass a `CoroutineScope` to the function that starts the work, such as `startLoading()` in this example, and launch the coroutine in that scope so callers can cancel the work when they no longer need it.
>
{style="note"}

> Since `StateFlow` is a hot flow, collection doesn't complete by itself.

Reviewer: ```suggestion
Since `StateFlow` is a hot flow,  the collection doesn’t complete on its own.
```

In this example, the `.takeWhile()` operator stops collection when loading reaches a terminal state.

A `StateFlow` emits an update only when the new value differs from the current value.

> Avoid storing mutable objects in a `StateFlow`.
> Mutating the object itself doesn't replace the current value, so collectors don't receive an update.
> 
{style="warning"}

You can also update a `StateFlow` by calculating the new state from the current state.
Use the [`.update()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/update.html) function for these updates.
The `.update()` function updates the value atomically, which helps when multiple coroutines update the same `MutableStateFlow`.

> If you only need to update a shared value and don't need to observe state changes over time, use the Kotlin Atomics APIs, such as [`AtomicInt`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.concurrent.atomics/-atomic-int/) or [`AtomicReference`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.concurrent.atomics/-atomic-reference/).
>
{style="note"}

> Here's an example where likes are stored in a `StateFlow` and each new state is calculated from the previous state:

Reviewer: ```suggestion
Here's an example where likes are stored in a `StateFlow`, and each new state is calculated from the previous state:
```

import kotlinx.coroutines.flow.*
import kotlin.time.*
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds
import kotlin.io.encoding.*
import java.io.IOException
import kotlin.random.Random

//sampleStart
class Post(val id: Long) {
    // Stores the current number of likes as a StateFlow
    private val _numberOfLikes = MutableStateFlow<Int>(
        // Sets the initial number of likes
        0
    )

    // Exposes a read-only StateFlow with the current number of likes
    val numberOfLikes: StateFlow<Int>
        get() = _numberOfLikes.asStateFlow()

    // Adds a like
    fun like() {
        // Increments the number of likes atomically for concurrent and multithreaded calls
        _numberOfLikes.update { it + 1 }
    }
}

suspend fun drawUpdatedNumberOfLikes(likes: Int) {
    // Displays the latest number of likes
    println("${Clock.System.now()}: the number of likes is $likes")
}

suspend fun main() {
    withContext(Dispatchers.Default) {
        val post = Post(15)
        val notifyingJob = launch {
            post.numberOfLikes.collect {
                drawUpdatedNumberOfLikes(it)
            }
        }

>         // Simulates users that like the post

Reviewer: ```suggestion
        // Simulates users who like the post
```

                    userId,
                    Clock.System.now(),
                    "Hello from $userId!"
                )
            )
        }
        // Delays to make sure users have enough time to receive updates
        delay(100.milliseconds)
        // Cancels readers because StateFlow collection doesn't finish by itself
        messageReaders.forEach { it.cancel() }
    }
}
//sampleEnd
```
{kotlin-runnable="true"}

In this example, `messageHistory` stores the full list of previous messages as the current state.
When a new message is sent, the `.update()` function creates a new list from the previous history and adds the new message atomically.

> Updating immutable collections by creating a new collection can take more time as the collection grows.
> You can create persistent collections with the [Experimental](components-stability.md#stability-levels-explained) [`kotlinx.collections.immutable`](https://github.com/Kotlin/kotlinx.collections.immutable) library to make immutable collection updates more efficient.
>
{style="tip"}

Since `messageHistory` is a `StateFlow`, subscribers receive the current message history when they start collecting.

> After that, they receive a new list each time a message is sent which changes the chat history.

Reviewer: ```suggestion
After that, they receive a new list each time a message is sent, which changes the chat history.
```

> ### Convert cold flows to hot flows

Reviewer: Do you place it here as another option to create hot flows? Maybe explicitly present it as "Create hot flows from cold flows?"

Author: That makes sense, and I agree that “create” is technically accurate. 🤔  I still prefer “convert” here because it better reflects the user’s goal iiuc: they start with a cold flow that is too expensive to collect repeatedly, then use `shareIn()` or `stateIn()` to convert it into a hot flow. Even though this creates a new hot flow, the important concept is the conversion from cold to hot.

Reviewer: Just from a page-structure point of view, if the main thing is creating hot flows from cold flows, it makes sense to keep it here under Hot flows, next to other "create" options. But if the focus is on the conversion, then I'd place it below on the second level since it is not a part of Hot flows or Cold flows:

```
## Cold flows
## Hot flows
## Convert
```

Author: as discussed on Slack let's keep the original structure then 👍

fun <T> Flow<T>.simpleShareIn(scope: CoroutineScope): SharedFlow<T> {
    val sharedFlow = MutableSharedFlow<T>()
    scope.launch {
        this@simpleShareIn.collect {
            sharedFlow.emit(it)
        }
    }
    return sharedFlow.asSharedFlow()
}

suspend fun main() { 
    
}
```

In this example, `simpleShareIn()` starts a new coroutine in the provided scope.
To stop collecting from the upstream flow, [cancel the scope](#cancel-hot-flows) that runs the collecting coroutine.

If the upstream flow throws an exception, this collecting coroutine fails.
Use operators such as `.catch()` or `.retry()` before sharing the flow to [handle upstream exceptions](#handle-exceptions-in-hot-flows) before the collecting coroutine fails.

The built-in [`.shareIn()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/share-in.html) function provides this pattern without requiring you to create a `MutableSharedFlow` yourself.
It also adds options for controlling when upstream collection starts and stops, and how many previous emissions new subscribers receive.

To use the built-in `.shareIn()` function, provide the following arguments:

> 1. The coroutine scope where the upstream flow is collected.

Reviewer: ```suggestion
* The coroutine scope where the upstream flow is collected.
```

> 2. The [`SharingStarted`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-sharing-started/) strategy to control when upstream collection starts and stops.

Reviewer: ```suggestion
* The [`SharingStarted`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-sharing-started/) strategy that controls when upstream collection starts and stops.
```

   For example, [`SharingStarted.Eagerly`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-sharing-started/-companion/-eagerly.html) starts upstream collection immediately in the provided scope, before any subscriber starts collecting.

> 3. The optional `replay` value to control how many previous emissions new subscribers receive.

Reviewer: ```suggestion
* The optional `replay` value that controls how many previous emissions new subscribers receive.
```

The `.shareIn()` function collects the upstream flow in the provided coroutine scope and broadcasts its emissions to subscribers.

Here's an example where `.shareIn()` converts a cold flow to a hot flow that shares serialized chat messages with multiple subscribers:

```kotlin
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import kotlin.time.*
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds
import kotlin.io.encoding.*
import java.io.IOException
import kotlin.random.Random

//sampleStart
data class Message(
    val senderId: Int,
    val time: Instant,
    val text: String,
)

class Chatroom {
    // Stores the message flow
    private val _messages = MutableSharedFlow<Message>()

    // Exposes a read-only SharedFlow with emitted messages
    // New subscribers don't receive already emitted messages
    val messages: SharedFlow<Message>
        get() = _messages.asSharedFlow()
  

>     // Sends a message to all subscribers of the messages flow

Reviewer: ```suggestion
    // Sends a message to all subscribers of the message flow
```

Author: same here, the flow is messages (individual value is message)

Reviewer: Yeah, I get it, but it's a standard noun as adjective (noun modifier), always singular, even if it means plural

Author: hmm 🤔 What I meant is that's the name of the flow: `messages` so calling it message is misleading (that's one value in the `messages` flow )

Like in another example // Sends a message to all subscribers of the messageHistory flow

Reviewer: oh, I didn't realize you're referring here to actual variable names! maybe add backticks?

Author: 🤔 
WDYT about like this:
    // Sends a message to all subscribers of the flow: messages
we never put backticks to comments for the other coroutines docs (I don't really see it in other docs either 🤔 ) 

or just leave it as is? 🤔

Because `SharingStarted.Eagerly` starts upstream collection immediately, the derived hot flow starts collecting `chatroom.messages` as soon as `.shareIn()` is called.

Similarly, to convert a cold flow to a `StateFlow`, use the [`.stateIn()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/state-in.html) function.

Unlike `.shareIn()`, `.stateIn()` requires an initial value because a `StateFlow` must always have a current value.

For example:

```kotlin
val lastUpdateFlow: StateFlow<Instant?> =
    chatroom
        .messageHistory
        .map { currentHistory -> currentHistory.lastOrNull()?.time }
        .stateIn(
            // Starts the sharing coroutine in this scope
            // The upstream flow, including .map(), runs in that coroutine
            derivedFlowsScope,

            // Starts collecting when the first subscriber appears
            // and stops when the last subscriber disappears
            SharingStarted.WhileSubscribed(),
            // Sets the initial state before the first upstream emission
            null,
        )
```

> #### Cancel hot flows

Reviewer: ```suggestion
### Cancel hot flows
```

Reviewer: Otherwise the header is not visible in the right-hand menu, which is inconsistent with Cold flows

Author: it is. We added it like this to try to ensure that people read these sections in the context of the 
Convert cold flows to hot flows section, because we were a bit worried that it might be confusing otherwise. 🤔 

Regardless, I agree it works as a 3rd level heading as well and makes navigation easier...

Hot flows don't stop when a subscriber is canceled.

When you cancel the coroutine that collects a hot flow, you only cancel that subscriber.
The hot flow can still emit values to other subscribers, and the coroutine that produces those values can keep running.

Hot flows themselves don't have a cancellation operation.
To cancel a hot flow, cancel the coroutine or scope that produces values for it.

> Hot flows created with the `.shareIn()` or `.stateIn()` extension functions keep collecting the upstream flow until the sharing coroutine is canceled.

Reviewer: ```suggestion
Hot flows created with the `.shareIn()` or `.stateIn()` extension functions continue to collect the upstream flow until the sharing coroutine is canceled.
```

> To stop collection from the upstream flow, cancel the scope that runs that sharing coroutine.

Reviewer: ```suggestion
To stop collection from the upstream flow, cancel the scope that runs the sharing coroutine.
```

> > You can also stop upstream collection automatically when there are no subscribers by using [`SharingStarted.WhileSubscribed()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-sharing-started/-companion/-while-subscribed.html).

Reviewer: ```suggestion
> You can also automatically stop upstream collection when there are no subscribers using [`SharingStarted.WhileSubscribed()`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-sharing-started/-companion/-while-subscribed.html).
```

Author: Both are correct, just a nit why I prefer the original: it contrasts the new idea more: You can stop X with this - you can also stop x automatically with that.

(agree with removing the by! 👍 🙏 )

                // Starts the sharing coroutine in this scope
                derivedFlowsScope
            )
        // Updates messageHistory
        chatroom.sendMessageToEveryone(
            Message(0, Clock.System.now(), "We are shutting down soon!")
        )
        delay(100.milliseconds)
        // Cancels the scope that runs the derived hot flow
        derivedFlowsScope.cancel()
        // Updates messageHistory, but totalMessages no longer receives the update
        chatroom.sendMessageToEveryone(
            Message(0, Clock.System.now(), "We have shut down.")
        )
        println("Last collected history size: ${totalMessages.value}")
        println("Actual history size: ${chatroom.messageHistory.value.size}")
    }
}
//sampleEnd
```
{kotlin-runnable="true"}

In this example, when you call the `derivedFlowsScope.cancel()` function, `totalMessages` stops collecting updates from `messageHistory`.

The `sendMessageToEveryone()` function still updates `messageHistory` because the coroutine that calls it wasn't canceled.
As a result, `totalMessages.value` keeps the last collected size, while `chatroom.messageHistory.value.size` shows the actual number of messages.

> #### Handle exceptions in hot flows

Reviewer: ```suggestion
### Handle exceptions in hot flows
```
